usr/opt/OSC200/usr/share/man/man3/gets.3
.\" This manpage source uses rsml coding. 
.so /usr/share/lib/tmac/sml
.so /usr/share/lib/tmac/rsml
.\"
.\"
.\" *********************************************************************
.\" *                                                                   *
.\" *      Copyright (c) Digital Equipment Corporation, 1991, 1994      *
.\" *                                                                   *
.\" *                       All Rights Reserved.                        *
.\" *                                                                   *
.\" *********************************************************************
.\"
.\"
.\" HISTORY
.\"
.\" @(#)$RCSfile: gets.3,v $ $Revision: 3.2.5.2 $ (DEC) $Date: 1993/06/24 18:54:00 $
.\"
.\" (c) Copyright 1990, 1991, 1992, 1993 OPEN SOFTWARE FOUNDATION, INC. 
.\" ALL RIGHTS RESERVED 
.\" OSF/1 1.2
.\"
.TH gets 3
.SH NAME
.PP
\*Lgets\*O, \*Lfgets\*O \- Get a string from a stream
.SH LIBRARY
.PP
Standard I/O Library (\*Llibc.a\*O)
.SH SYNOPSIS
.PP
.sS
\*L#include <stdio.h>\*O
.PP
\*Lchar \(**gets(
.\" .zA "defect, 7508, R1.1.2, Removed erroneous const"
.nL
        \*Lchar \(**\*Vstring\*L);\*O
.\" .zZ "defect, 7508, R1.1.2, Removed erroneous const"
.PP
\*Lchar *fgets(
.\" .zA "defect, 7508, R1.1.2, Removed erroneous const"
.nL
        char *\*Vstring\*L,
.\" .zZ "defect, 7508, R1.1.2, Removed erroneous const"
.nL
        int \*Vn\*L,
.nL
        FILE *\*Vstream\*L);\*O
.sE
.SH PARAMETERS
.PP
.VL
.LI "\*Vstring\*O"
Points to a string to receive bytes.
.LI "\*Vn\*O"
Specifies an upper bound on the number of bytes to read.
.LI "\*Vstream\*O"
Points to the \*LFILE\*O structure of an open file.
.LE
.SH DESCRIPTION
.PP
.iX "gets function"
.iX "fgets function"
.iX "stream" "getting a string from stdin"
.iX "string" "getting from a stream"
The \*Lgets(\|)\*O function reads bytes from the standard
input stream, \*Lstdin\*O, into the array pointed to by the
\*Vstring\*O parameter. Data is read until a newline character is read
or an End-of-File condition is encountered. If reading is stopped
due to a newline character, the newline character is discarded and
the string is terminated with a NULL character.
.PP
The \*Lfgets(\|)\*O function reads bytes from the data pointed
to by the \*Vstream\*O parameter into the array pointed to by the \*Vstring\*O
parameter. Data is read until \*Vn\*O-1
bytes have been read, until a newline character is read and
transferred to \*Vstring\*O, or until an EOF (End-of-File) condition
is encountered. The string is then terminated with a NULL character.
.SH NOTES
.PP
.VL 1.5i
.LI "\*LAES Support Level:\*O"
Full use.
.LE
.SH CAUTIONS
.PP
The \*Lgets(\|)\*O function does not check the input for a maximum
size. Consequently, if more bytes are entered than will fit in
the space allocated for the \*Vstring\*O parameter, \*Lgets(\|)\*O
will write beyond the end of the allocated space, producing
indeterminate results. To avoid this condition, use
\*Lfgets(\|)\*O instead of \*Lgets(\|)\*O.
.SH RETURN VALUES
.PP
If the end of the file is encountered and no characters have been
read, no characters are transferred to \*Vstring\*O and a NULL
pointer is returned. If a read error occurs, a NULL pointer
is returned. Otherwise, \*Vstring\*O is returned.
.SH RELATED INFORMATION
.PP
Functions:
\*Lclearerr\*O(3),
\*Lfeof\*O(3),
\*Lferror\*O(3),
\*Lfgetws\*O(3),
\*Lfileno\*O(3),
\*Lfopen\*O(3),
\*Lfputws\*O(3),
\*Lfread\*O(3),
\*Lgetc\*O(3),
\*Lgetwc\*O(3),
\*Lputs\*O(3),
\*Lscanf\*O(3).
.EQ
delim off
.EN


usr/opt/OSC200/src/usr/ccs/lib/libc/gets.c
/*
 * *****************************************************************
 * *                                                               *
 * *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
 * *                                                               *
 * *   All Rights Reserved.  Unpublished rights  reserved  under   *
 * *   the copyright laws of the United States.                    *
 * *                                                               *
 * *   The software contained on this media  is  proprietary  to   *
 * *   and  embodies  the  confidential  technology  of  Digital   *
 * *   Equipment Corporation.  Possession, use,  duplication  or   *
 * *   dissemination of the software and media is authorized only  *
 * *   pursuant to a valid written license from Digital Equipment  *
 * *   Corporation.                                                *
 * *                                                               *
 * *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
 * *   by the U.S. Government is subject to restrictions  as  set  *
 * *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
 * *   or  in  FAR 52.227-19, as applicable.                       *
 * *                                                               *
 * *****************************************************************
 */
/*
 * HISTORY
 */
#ifndef lint
static char	*sccsid = "@(#)$RCSfile: gets.c,v $ $Revision: 4.3.8.2 $ (DEC) $Date: 1993/06/07 23:02:34 $";
#endif 
/*
 * (c) Copyright 1990, 1991, 1992, 1993 OPEN SOFTWARE FOUNDATION, INC. 
 * ALL RIGHTS RESERVED 
 */
/*
 * OSF/1 1.2
 */
#if !defined(lint) && !defined(_NOIDENT)

#endif
/*
 * FUNCTIONS: gets 
 *
 * (C) COPYRIGHT International Business Machines Corp. 1985, 1989 
 * All Rights Reserved
 *
 * US Government Users Restricted Rights - Use, duplication or
 * disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
 *
 * gets.c	1.8  com/lib/c/io,3.1,8943 9/12/89 18:32:40
 */

/*LINTLIBRARY*/
/*
 * This version reads directly from the buffer rather than looping on getc.
 * Ptr args aren't checked for NULL because the program would be a
 * catastrophic mess anyway.  Better to abort than just to return NULL.
 */

/* name space pollution clean up */
#ifdef _NAME_SPACE_WEAK_STRONG
#include "pollution.h"
#endif
#include <errno.h>
#include <stdio.h>
#include "stdiom.h"
#include "ts_supp.h"

#ifdef	_THREAD_SAFE
#include "stdio_lock.h"
#endif	/* _THREAD_SAFE */

extern int __filbuf();
extern char *memccpy();

char *
gets(char *s)
{
	char	*p, *s0 = s;
	int	n;
	TS_FDECLARELOCK(filelock)

	TS_FLOCK(filelock, stdin);
	if ((stdin->_flag&(_IOREAD|_IORW))==0) {  /* not open for reading */
                TS_SETERR(EBADF);
	        TS_FUNLOCK(filelock);
                return(NULL);
	}

	for ( ; ; ) {
		if (stdin->_cnt <= 0) { /* empty buffer */
			if (__filbuf(stdin) == EOF) {
				if (s0 == s) {
	                                TS_FUNLOCK(filelock);
					return (NULL);
                                }
				break; /* no more data */
			}
			stdin->_ptr--;
			stdin->_cnt++;
		}
		n = stdin->_cnt;
		if ((p = memccpy(s, (char *) stdin->_ptr, '\n', n)) != NULL)
			n = p - s;
		s += n;
		stdin->_cnt -= n;
		stdin->_ptr += n;
		_BUFSYNC(stdin);
		if (p != NULL) { /* found '\n' in buffer */
			s--; /* step back over '\n' */
			break;
		}
	}
	*s = '\0';
        TS_FUNLOCK(filelock);
	return (s0);
}
